core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,394 @@
+	}
+
+	/* Perform a dummy read to initiate the receive operation. */
+	io_read8((vaddr_t)&regs->ibdr);
+
+	for (n = 0; n < operation->length_in_bytes; n++) {
+		res = i2c_transfer_complete(regs, I2C_BUS_NO_TEST_RX_ACK);
+		if (res)
+			return res;
+		if (n == (operation->length_in_bytes - 2)) {
+			/* Set No ACK = 1 */
+			io_setbits8((vaddr_t)&regs->ibcr, I2C_IBCR_NOACK);
+		} else if (n == (operation->length_in_bytes - 1)) {
+			if (!is_last_operation) {
+				/* select Transmit mode (for repeat start) */
+				io_setbits8((vaddr_t)&regs->ibcr,
+					    I2C_IBCR_TXRX);
+			} else {
+				/* Generate Stop Signal */
+				io_clrbits8((vaddr_t)&regs->ibcr,
+					    (I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+				res = i2c_bus_test_bus_busy(regs,
+							    I2C_BUS_TEST_IDLE);
+				if (res)
+					return res;
+			}
+		}
+		operation->buffer[n] = io_read8((vaddr_t)&regs->ibdr);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_write(struct i2c_regs *regs, unsigned int slave_address,
+			    struct i2c_operation *operation)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int n = 0;
+
+	/* Write Slave Address */
+	io_write8((vaddr_t)&regs->ibdr,
+		  (slave_address << 0x1) & ~(BIT(0)));
+	res = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (res)
+		return res;
+
+	/* Write Data */
+	for (n = 0; n < operation->length_in_bytes; n++) {
+		io_write8((vaddr_t)&regs->ibdr, operation->buffer[n]);
+		res = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_stop(struct i2c_regs *regs)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t reg = 0;
+
+	reg = io_read8((vaddr_t)&regs->ibsr);
+	if (reg & I2C_IBSR_IBB) {
+		/* Generate Stop Signal */
+		io_clrbits8((vaddr_t)&regs->ibcr,
+			    I2C_IBCR_MSSL | I2C_IBCR_TXRX);
+		res = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_IDLE);
+		if (res)
+			return res;
+	}
+
+	/* Disable I2c Controller */
+	io_setbits8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_start(struct i2c_regs *regs)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	io_setbits8((vaddr_t)&regs->ibsr, I2C_IBSR_IBAL | I2C_IBSR_IBIF);
+	io_clrbits8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+
+	/* Wait controller to be stable */
+	mdelay(1);
+
+	/* Generate Start Signal */
+	io_setbits8((vaddr_t)&regs->ibcr, I2C_IBCR_MSSL);
+	res = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_BUSY);
+	if (res)
+		return res;
+
+	/* Select Transmit Mode. set No ACK = 1 */
+	io_setbits8((vaddr_t)&regs->ibcr, I2C_IBCR_TXRX | I2C_IBCR_NOACK);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Transfer data to/from I2c slave device
+ * base           Base Address of I2c controller's registers
+ * slave_address  Slave Address from which data is to be read
+ * request_packet Pointer to an i2c_reg_request structure
+ * describing the I2C transaction
+ */
+TEE_Result i2c_bus_xfer(vaddr_t base, unsigned int slave_address,
+			struct i2c_reg_request *request_packet)
```

Comment:
It seems that the `struct i2c_reg_request` is only used to pass `operation_count` and `operation` to this function.
Base on what I can see in this patch you might as well take those as separate parameters and things would be a little bit simpler.

Reply:
Yeah, sure will update.